sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/FluxAutoComplete.java

```diff
@@ -0,0 +1,130 @@
+ */
+class FluxAutoComplete<T extends ServiceBusReceivedMessageContext> extends FluxOperator<T, T> {
+    private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete;
+    private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon;
+    private final ClientLogger logger = new ClientLogger(FluxAutoComplete.class);
+
+    FluxAutoComplete(Flux<? extends T> upstream, Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete,
+        Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon) {
+        super(upstream);
+        this.onComplete = Objects.requireNonNull(onComplete, "'onComplete' cannot be null.");
+        this.onAbandon = Objects.requireNonNull(onAbandon, "'onAbandon' cannot be null.");
+    }
+
+    /**
+     * Invoked when a downstream subscriber is interested in objects published from this operator.
+     *
+     * @param coreSubscriber The subscriber interested in the published items.
+     */
+    @Override
+    public void subscribe(CoreSubscriber<? super T> coreSubscriber) {
+        Objects.requireNonNull(coreSubscriber, "'coreSubscriber' cannot be null.");
+
+        final AutoCompleteSubscriber<T> subscriber =
+            new AutoCompleteSubscriber<T>(coreSubscriber, onComplete, onAbandon, logger);
+
+        source.subscribe(subscriber);
+    }
+
+    static final class AutoCompleteSubscriber<T extends ServiceBusReceivedMessageContext> extends BaseSubscriber<T> {
+        private final CoreSubscriber<? super T> downstream;
+        private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete;
+        private final Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon;
+        private final ClientLogger logger;
+
+        AutoCompleteSubscriber(CoreSubscriber<? super T> downstream,
+            Function<ServiceBusReceivedMessageContext, Mono<Void>> onComplete,
+            Function<ServiceBusReceivedMessageContext, Mono<Void>> onAbandon, ClientLogger logger) {
+            this.downstream = downstream;
+            this.onComplete = onComplete;
+            this.onAbandon = onAbandon;
+            this.logger = logger;
+        }
+
+        @Override
+        protected void hookOnSubscribe(Subscription subscription) {
+            logger.info("Subscription received. Subscribing downstream. {}", subscription);
+            downstream.onSubscribe(this);
+            requestUnbounded();
+        }
+
+        @Override
+        protected void hookOnNext(T value) {
+            final ServiceBusReceivedMessage message = value.getMessage();
+            final String sequenceNumber = message != null ? String.valueOf(message.getSequenceNumber()) : "n/a";
+
+            logger.verbose("Passing message downstream. sequenceNumber[{}]", sequenceNumber);
+
+            try {
+                downstream.onNext(value);
+                applyWithCatch(onComplete, value, "complete");
```

Comment:
This doesn't have to be called if the message is `null`. Same for abandon too.

Reply:
If complete fails , there is chance that message will be processed twice because it will go back to queue. 
For now, we should log and continue ?
But as part of for-ever loop, this should me delivered to user in their error handler with indication that we fail to complete.

Reply:
I separated that logic into the responsibility of the async client.

```java
            return new FluxAutoComplete(messageFlux,
                context -> context.getMessage() != null ? complete(context.getMessage()) : Mono.empty(),
                context -> context.getMessage() != null ? abandon(context.getMessage()) : Mono.empty());
```